test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestValueRematDuringTypeSharpening.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.valhalla.inlinetypes;
+
+import jdk.test.lib.Asserts;
+import jdk.internal.misc.Unsafe;
+import compiler.lib.ir_framework.*;
+
+/*
+ * @test
+ * @summary Missing InlineTypeNode re-materialization during type sharpening.
+ * @library /test/lib /
+ * @compile --add-exports java.base/jdk.internal.misc=ALL-UNNAMED TestValueRematDuringTypeSharpening.java
+ * @run main/othervm/timeout=300 --add-exports java.base/jdk.internal.misc=ALL-UNNAMED compiler.valhalla.inlinetypes.TestValueRematDuringTypeSharpening
+ */
+
+abstract class topValue {
+}
+
+value class dummyValue1 extends topValue {
+   int field;
+   public dummyValue1(int val) {
+      field = val;
+   }
+}
+
+value class dummyValue2 extends topValue {
+   int field;
+   public dummyValue2(int val) {
+      field = val;
+   }
+}
+
+public class TestValueRematDuringTypeSharpening {
+
+   public static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+   @DontInline
+   public static int getUnsafeFieldValue(topValue obj, int incr) {
+       return UNSAFE.getInt(obj, 12) + incr;
+   }
+
+   @Test
+   @IR(phase = {CompilePhase.AFTER_PARSING}, counts = {IRNode.INLINE_TYPE, " > 0 "})
+   public static int test(topValue obj) {
+       int val = 0;
+       if (obj.getClass() != dummyValue1.class) {
+           val += 10;
+       } else if (obj.getClass() != dummyValue2.class) {
+           val += 20;
+       }
+       return getUnsafeFieldValue(obj, val);
+   }
+
+   @Run(test = {"test"}, mode = RunMode.STANDALONE)
+   public static void kernel() {
+       int res = 0;
+       for (int i = 0; i < 10000; i++) {
+           res += test((i % 1000) == 1 ? new dummyValue1(i) : new dummyValue2(i));
+       }
```

Comment:
I think you should leave warmup to the IR framework. If more iterations are needed, you can control that via the `@Warmup` annotation.

Reply:
Thanks @TobiHartmann , standalone run mode has no warmups hence there was an explicit warmup loop, changed as per your suggestion.